File: web/src/lib/pageMetricsCompare.ts
import type { CompareMetricRow } from '@/lib/reportCompare';
import type { PageGa4Slice, PageGscSlice } from '@/server/pageGoogleData';

function num(v: unknown): number | null {
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

function deltaRow(
  id: string,
  label: string,
  current: number | null,
  baseline: number | null,
  higherIsBetter: boolean,
  format: CompareMetricRow['format'] = 'count',
): CompareMetricRow {
  const delta =
    current != null && baseline != null ? Math.round((current - baseline) * 10) / 10 : null;
  let deltaPct: number | null = null;
  if (current != null && baseline != null && baseline !== 0) {
    deltaPct = Math.round(((current - baseline) / Math.abs(baseline)) * 1000) / 10;
  }
  return { id, label, current, baseline, delta, higherIsBetter, format, deltaPct };
}

export type PageMetricsPayload = {
  gsc?: PageGscSlice | null;
  ga4?: PageGa4Slice | null;
};

export function buildPageMetricsCompare(
  current: PageMetricsPayload,
  baseline: PageMetricsPayload,
  labels: {
    gscClicks: string;
    gscImpressions: string;
    gscCtr: string;
    gscPosition: string;
    ga4Sessions: string;
    ga4Users: string;
    ga4Views: string;
    ga4Engagement: string;
    ga4Duration: string;
  },
): CompareMetricRow[] {
  const cg = current.gsc;
  const bg = baseline.gsc;
  const ca = current.ga4;
  const ba = baseline.ga4;
  const rows: CompareMetricRow[] = [];

  if (cg || bg) {
    rows.push(
      deltaRow('gsc_clicks', labels.gscClicks, num(cg?.clicks), num(bg?.clicks), true),
      deltaRow('gsc_impr', labels.gscImpressions, num(cg?.impressions), num(bg?.impressions), true),
      deltaRow('gsc_ctr', labels.gscCtr, num(cg?.ctr), num(bg?.ctr), true, 'percent'),
      deltaRow('gsc_pos', labels.gscPosition, num(cg?.position), num(bg?.position), false),
    );
  }
  if (ca || ba) {
    rows.push(
      deltaRow('ga4_sessions', labels.ga4Sessions, num(ca?.sessions), num(ba?.sessions), true),
      deltaRow('ga4_users', labels.ga4Users, num(ca?.activeUsers), num(ba?.activeUsers), true),
      deltaRow(
        'ga4_views',
        labels.ga4Views,
        num(ca?.screenPageViews),
        num(ba?.screenPageViews),
        true,
      ),
      deltaRow(
        'ga4_engagement',
        labels.ga4Engagement,
        num(ca?.engagementRate),
        num(ba?.engagementRate),
        true,
        'percent',
      ),
      deltaRow(
        'ga4_duration',
        labels.ga4Duration,
        num(ca?.avgSessionDuration),
        num(ba?.avgSessionDuration),
        true,
      ),
    );
  }
  return rows.filter((r) => r.current != null || r.baseline != null);
}

File: services/CoreService/src/CoreService.Api/IntegrationsApplication/Google/PageMetricsCompare.cs
namespace CoreService.Api.IntegrationsApplication.Google;

/// <summary>
/// Parity with web/src/lib/pageMetricsCompare.ts — page-level GSC/GA4 compare rows.
/// </summary>
public static class PageMetricsCompare
{
    public static IReadOnlyList<PageCompareMetricRow> Build(
        PageMetricsPayload current,
        PageMetricsPayload baseline)
    {
        var rows = new List<PageCompareMetricRow>();
        if (current.Gsc is not null || baseline.Gsc is not null)
        {
            rows.Add(DeltaRow("gsc_clicks", "Clicks", current.Gsc?.Clicks, baseline.Gsc?.Clicks, higherIsBetter: true));
            rows.Add(DeltaRow("gsc_impr", "Impressions", current.Gsc?.Impressions, baseline.Gsc?.Impressions, higherIsBetter: true));
            rows.Add(DeltaRow("gsc_ctr", "CTR %", current.Gsc?.Ctr, baseline.Gsc?.Ctr, higherIsBetter: true, format: "percent"));
            rows.Add(DeltaRow("gsc_pos", "Avg position", current.Gsc?.Position, baseline.Gsc?.Position, higherIsBetter: false));
        }

        if (current.Ga4 is not null || baseline.Ga4 is not null)
        {
            rows.Add(DeltaRow("ga4_sessions", "Sessions", current.Ga4?.Sessions, baseline.Ga4?.Sessions, higherIsBetter: true));
            rows.Add(DeltaRow("ga4_users", "Users", current.Ga4?.ActiveUsers, baseline.Ga4?.ActiveUsers, higherIsBetter: true));
            rows.Add(DeltaRow("ga4_views", "Page views", current.Ga4?.ScreenPageViews, baseline.Ga4?.ScreenPageViews, higherIsBetter: true));
            rows.Add(DeltaRow("ga4_engagement", "Engagement rate", current.Ga4?.EngagementRate, baseline.Ga4?.EngagementRate, higherIsBetter: true, format: "percent"));
            rows.Add(DeltaRow("ga4_duration", "Avg session (s)", current.Ga4?.AvgSessionDuration, baseline.Ga4?.AvgSessionDuration, higherIsBetter: true));
        }

        return rows.Where(r => r.Current is not null || r.Baseline is not null).ToList();
    }

    private static PageCompareMetricRow DeltaRow(
        string id,
        string label,
        double? current,
        double? baseline,
        bool higherIsBetter,
        string format = "count")
    {
        double? delta = current is not null && baseline is not null
            ? Math.Round(current.Value - baseline.Value, 1)
            : null;
        double? deltaPct = null;
        if (current is not null && baseline is not null && baseline.Value != 0)
        {
            deltaPct = Math.Round((current.Value - baseline.Value) / Math.Abs(baseline.Value) * 1000) / 10;
        }

        return new PageCompareMetricRow
        {
            Id = id,
            Label = label,
            Current = current,
            Baseline = baseline,
            Delta = delta,
            HigherIsBetter = higherIsBetter,
            Format = format,
            DeltaPct = deltaPct,
        };
    }
}

public sealed class PageMetricsPayload
{
    public PageGscMetrics? Gsc { get; init; }

    public PageGa4Metrics? Ga4 { get; init; }
}

public sealed class PageGscMetrics
{
    public double? Clicks { get; init; }

    public double? Impressions { get; init; }

    public double? Ctr { get; init; }

    public double? Position { get; init; }
}

public sealed class PageGa4Metrics
{
    public double? Sessions { get; init; }

    public double? ActiveUsers { get; init; }

    public double? ScreenPageViews { get; init; }

    public double? EngagementRate { get; init; }

    public double? AvgSessionDuration { get; init; }
}

public sealed class PageCompareMetricRow
{
    public string Id { get; init; } = "";

    public string Label { get; init; } = "";

    public double? Current { get; init; }

    public double? Baseline { get; init; }

    public double? Delta { get; init; }

    public bool HigherIsBetter { get; init; }

    public string Format { get; init; } = "count";

    public double? DeltaPct { get; init; }
}

public sealed class PageCompareArm
{
    public string Type { get; init; } = "snapshot";

    public long Id { get; init; }

    public string? FetchedAt { get; init; }

    public object? Gsc { get; init; }

    public object? Ga4 { get; init; }

    public PageMetricsPayload ToMetricsPayload() =>
        new()
        {
            Gsc = PageCompareMetricsParser.FromGscObject(Gsc),
            Ga4 = PageCompareMetricsParser.FromGa4Object(Ga4),
        };
}

public static class PageCompareMetricsParser
{
    public static PageGscMetrics? FromGscObject(object? gsc) =>
        gsc is Dictionary<string, object?> dict
            ? new PageGscMetrics
            {
                Clicks = ReadDouble(dict, "clicks"),
                Impressions = ReadDouble(dict, "impressions"),
                Ctr = ReadDouble(dict, "ctr"),
                Position = ReadDouble(dict, "position"),
            }
            : null;

    public static PageGa4Metrics? FromGa4Object(object? ga4) =>
        ga4 is Dictionary<string, object?> dict
            ? new PageGa4Metrics
            {
                Sessions = ReadDouble(dict, "sessions"),
                ActiveUsers = ReadDouble(dict, "activeUsers") ?? ReadDouble(dict, "active_users"),
                ScreenPageViews = ReadDouble(dict, "screenPageViews") ?? ReadDouble(dict, "screen_page_views"),
                EngagementRate = ReadDouble(dict, "engagementRate") ?? ReadDouble(dict, "engagement_rate"),
                AvgSessionDuration = ReadDouble(dict, "avgSessionDuration") ?? ReadDouble(dict, "avg_session_duration"),
            }
            : null;

    private static double? ReadDouble(Dictionary<string, object?> dict, string key)
    {
        if (!dict.TryGetValue(key, out var raw) || raw is null)
        {
            return null;
        }

        return raw switch
        {
            double d => d,
            float f => f,
            int i => i,
            long l => l,
            decimal m => (double)m,
            string s when double.TryParse(s, out var parsed) => parsed,
            _ => null,
        };
    }
}
